### PR TITLE
Implement threshold cryptography

### PR DESCRIPTION
**Pull request by abread**
This PR introduces `threshcrypto` - a new module for working with threshold signatures in Mir, and an implementation using BLS12-381 threshold signatures (using the [same underlying library as drand](https://github.com/drand/kyber)).

There was [some discussion on Slack](https://filecoinproject.slack.com/archives/C03C77HN3AS/p1660755716337459) on how to go about this.

It's based on the existing `crypto` module: the dummy and pseudo implementations were kept for testing purposes, but I did not reimplement the `DefaultImpl` as no API consumers exist *yet* and there's no "real" implementations besides TBLS to guide its design.

### CODE REVIEW

File: pkg/threshcrypto/api.go
Comment (xosmig) [inline pkg/threshcrypto/api.go:27]: I have mixed feelings regarding this property.
The implementation you are using does indeed verify the partial signatures in `Recover`.
However, it seems kind of redundant in my opinion as the shares can (and, probably, should) be verified before invoking `Recover`.

Maybe let's make it a property of the particular implementation and not of the abstraction (i.e., move it to the comment for `TBLSInst.Recover`)?

**Review by xosmig — COMMENTED** [review]
Thanks a lot for the PR! I left a few minor cosmetic comments and a couple of comments w.r.t. to the interface.

File: pkg/threshcrypto/api.go
Comment (xosmig) [reply pkg/threshcrypto/api.go:27]: In fact, it could be a serious performance overhead as verifying signature shares (i.e., doing pairings) is the most computationally heavy part. Not only this API forces the protocol to verify the shares multiple times (once before `Recover` and once inside), but it also prohibits parallelizing the verification.

I wonder why they did it this way.

Comment (abread) [reply pkg/threshcrypto/api.go:27]: I agree, it's a good idea to force the optimizable design (verify shares, then recover) right from the start, even if some implementations are sub-optimal.

The implementation of `Recover` for tbls only uses public kyber APIs and is simple, so I think we can land the optimized implementation (without the double verification) here.

File: pkg/threshcrypto/api.go
Comment (abread) [reply pkg/threshcrypto/api.go:23]: For BLS it doesn't matter a lot: signatures (and sig shares) are unique and deterministic, so I think it's equivalent to wait for `t` distinct valid shares before calling `Recover`.

That said, other schemes may benefit from this, and some protocols may want to classify a mismatch as byzantine behavior, so I agree that we should include it.

**Comment by abread** [comment]
@xosmig Thank you for the thorough review! I believe I have integrated all the suggestions

File: pkg/threshcrypto/tbls_test.go
Comment (matejpavlovic) [inline pkg/threshcrypto/tbls_test.go:124]: It would be cleaner to use `types.NodeIDFromInt` here.

**Comment by abread** [comment]
Thank you so much, Matej!
I've updated the PR with your suggestions (and preallocated an array as suggested by the linter)

File: go.mod
Comment (dnkolegov) [inline go.mod:30]: Is it safe to use this library? Why not use blst bindings?

**Review by xosmig — APPROVED** [review]
Thanks a lot!
**UPD:** I added a few more comments. Most are trivial, but the question about the minimum threshold is important.

File: pkg/threshcrypto/tbls_test.go
Comment (xosmig) [inline pkg/threshcrypto/tbls_test.go:120]: seed is unused

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:52]: Could you give a reference for this? I believe many applications need (t+1, 3t+1) threshold signatures.

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:81]:
```suggestion
// MarshalTo writes the properties of a TBLSInst to an io.Writer.
// Can be read with TBLSInst.UnmarshalFrom.
```

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:257]:
```suggestion
// SignShare constructs a signature share for the message.
```

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:262]:
```suggestion
// VerifyShare verifies that a signature share is for a given message from a given node.
```

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:276]:
```suggestion
// VerifyFull verifies that a (full) signature is valid for a given message.
```

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:282]:
```suggestion
// Recover recovers a full signature from a set of (previously validated) shares, that are known to be from
// distinct nodes.
```

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [inline pkg/threshcrypto/tbls.go:157]:
```suggestion
// UnmarshalFrom sets the properties of a TBLSInst from an io.Reader.
// The property stream can be created from TBLSInst.MarshalTo.
// NOTE: Currently assumes the underlying scheme is the same as in TBLS12381Keygen().
```

File: pkg/threshcrypto/tbls.go
Comment (abread) [reply pkg/threshcrypto/tbls.go:52]: Ah oops, I had a "TODO" comment precisely to remember to review this check carefully, and must have accidentally deleted it at some point

Comment (abread) [reply pkg/threshcrypto/tbls.go]: Ok, I reviewed the [literature](https://link.springer.com/content/pdf/10.1007/3-540-36288-6_3.pdf), and comitted a correct version, that supports the (t+1, 3t+1) use case.

It's hard (impossible?) to dissociate the check from the number of malicious nodes.

Comment (xosmig) [reply pkg/threshcrypto/tbls.go]: I would suggest removing this check altogether as it is rather trivial and concerns rather the protocol using the threshold signatures than the threshold signatures themselves.

I.e., you can have a (2, n)-threshold signature scheme, why not. It's just that, in the classical BFT model, you are unlikely to find such a signature useful, but it is not a concern of the threshold signature module.

Comment (xosmig) [reply pkg/threshcrypto/tbls.go]: Or am I missing something?
Where does `nByz < (N+1)/2` come from?

Comment (xosmig) [reply pkg/threshcrypto/tbls.go]: My speculation is that this is the requirement not for the threshold signature scheme itself, but for the interactive signing protocol. However, I need to check.

**Comment by abread** [comment]
I've started playing around with this, and ended up implementing some DSL utilities akin to the availability module.

They don't require much thought to implement so I went ahead and included them here

File: pkg/threshcrypto/tbls.go
Comment (xosmig) [reply pkg/threshcrypto/tbls.go]: Sorry, for taking so much time. I believe that the `t < n/2` part in the paper comes from the way the threshold signature scheme is used (in particular, the DKG part and the protocol for signing a message), which is kind of irrelevant.
So, I stick to my suggestion to remove this check completely.

**Comment by xosmig** [comment]
@abread sorry that you had to do it several times already, but could you please merge the latest `main` branch into your branch?
Also, would it be OK if we use the "squash and merge" option? (in the main branch, there will be just 1 `"Implement threshold cryptography"` commit, which will include all the changes from this PR). Otherwise, you would need to clean up the commit history and rebase it over the current `main` (as opposed to merging `main` into your branch).

**Comment by abread** [comment]
> sorry that you had to do it several times already, but could you please merge the latest main branch into your branch?
Ah, it was already done (when I added the DSL stuff) :)

> Also, would it be OK if we use the "squash and merge" option? 
Fine by me!

**Comment by xosmig** [comment]
> Ah, it was already done (when I added the DSL stuff) :)

Apparently, you will need to do it again.

image

**Comment by abread** [comment]
Weird, it gives me "Already up to date".
I'll try to rebase it against `main`

**Comment by abread** [comment]
Rebased to main, should be alright now!

**Comment by xosmig** [comment]
> Weird, it gives me "Already up to date".

Yes, sorry, it was my fault, I had the wrong option selected in github and the button for merging was gray 😩 
I removed my comment after I figured that out, but, apparently, too late.
I'm just waiting for the final OK from @matejpavlovic before merging.

**Comment by abread** [comment]
Ah dw, it was quick to do!

**Comment by matejpavlovic** [comment]
Merging. Thanks for your contribution!!

**Pull request merged**
